File: android/app/src/main/java/com/tvd12/freechat/socket/SocketRequests.kt
package com.tvd12.freechat.socket

import com.tvd12.ezyfoxserver.client.EzyClient
import com.tvd12.ezyfoxserver.client.EzyClients
import com.tvd12.ezyfoxserver.client.entity.EzyApp
import com.tvd12.ezyfoxserver.client.factory.EzyEntityFactory

import com.tvd12.freechat.constant.Commands
import com.tvd12.freechat.request.GetContactsRequest
import com.tvd12.freechat.request.SendSystemMessageRequest
import com.tvd12.freechat.request.SendUserMessageRequest

class SocketRequests private constructor() {
    companion object {
        fun sendGetContacts() {
            getApp()?.send(GetContactsRequest(0, 50))
        }

        fun sendGetSuggestContacts() {
            getApp()?.send(Commands.SUGGEST_CONTACTS)
        }

        fun sendAddContact(user: String) {
            sendAddContacts(listOf(user))
        }

        fun sendSearchContacts(keyword: String) {
            val app = getApp()
            if(app != null) {
                val data = EzyEntityFactory.newObjectBuilder()
                    .append("keyword", keyword)
                    .build()
                app.send(Commands.SEARCH_CONTACTS, data)
            }
        }

        private fun sendAddContacts(users: Collection<String>) {
            val app = getApp()
            if(app != null) {
                val data = EzyEntityFactory.newObjectBuilder()
                    .append("target", users)
                    .build()
                app.send(Commands.ADD_CONTACTS, data)
            }
        }

        fun sendSystemMessage(message: String) {
            getApp()?.send(SendSystemMessageRequest(message))
        }

        fun sendUserMessage(channelId: Long, message: String) {
            getApp()?.send(SendUserMessageRequest(channelId, message))
        }

        private fun getApp(): EzyApp? {
            val zone = getClient()?.zone ?: return null
            return zone.app
        }

        private fun getClient(): EzyClient? {
            val clients = EzyClients.getInstance()
            return clients.defaultClient
        }
    }
}

File: swift/freechat-swift/SocketRequests.swift
//
//  SocketRequests.swift
//  freechat-swift
//
//  Created by Dzung on 10/05/2021.
//  Copyright © 2021 Young Monkeys. All rights reserved.
//

public class SocketRequests {
    private init() {}
    
    public static func sendGetContacts() {
        let requestData = NSMutableDictionary()
        requestData["skip"] = 0
        requestData["limit"] = 100
        getApp()?.send(cmd: Commands.GET_CONTACTS, data: requestData)
    }
    
    public static func sendGetSuggestContacts() {
        getApp()?.send(cmd: Commands.SUGGEST_CONTACTS)
    }
    
    public static func sendBotMessage(message: String) {
        let requestData = NSMutableDictionary()
        requestData["message"] = message
        getApp()?.send(cmd: Commands.CHAT_BOT_MESSAGE, data: requestData)
    }
    
    public static func sendUserMessage(channelId: Int64, message: String) {
        let requestData = NSMutableDictionary()
        requestData["channelId"] = channelId
        requestData["message"] = message
        getApp()?.send(cmd: Commands.CHAT_USER_MESSAGE, data: requestData)
    }
    
    public static func sendSearchContacts(keyword: String) {
        let requestData = NSMutableDictionary()
        requestData["keyword"] = keyword
        getApp()?.send(cmd: Commands.SEARCH_CONTACTS, data: requestData)
    }
    
    public static func sendSearchExistedContacts(keyword: String) {
        let requestData = NSMutableDictionary()
        requestData["skip"] = 0
        requestData["limit"] = 100
        requestData["keyword"] = keyword
        getApp()?.send(cmd: Commands.SEARCH_EXISTED_CONTACTS, data: requestData);
    }
    
    public static func sendAddContacts(user: String) {
        let requestData = NSMutableDictionary()
        requestData["target"] = [user]
        getApp()?.send(cmd: Commands.ADD_CONTACTS, data: requestData)
     }
    
    private static func getApp() -> EzyApp? {
        return EzyClients.getInstance()
            .getDefaultClient()
            .getApp()
    }
}
